plugin/references.py

```diff
@@ -113,17 +130,27 @@ def _handle_response(
         force_group: bool,
         fallback: bool,
         group: int,
+        output_mode: Optional[OutputMode],
+        event: Optional[dict],
         position: int,
         response: Optional[List[Location]]
     ) -> None:
-        if response:
-            if userprefs().show_references_in_quick_panel:
-                self._show_references_in_quick_panel(
-                    word, session, response, side_by_side, force_group, group, position)
-            else:
-                self._show_references_in_output_panel(word, session, response)
-        else:
+        if not response:
             self._handle_no_results(fallback, side_by_side)
+            return
+        modifier_keys = (event or {}).get('modifier_keys', {})
+        if output_mode is None:
+            show_in_quick_panel = userprefs().show_references_in_quick_panel
+            if modifier_keys.get('shift'):
+                show_in_quick_panel = not show_in_quick_panel
+        else:
+            show_in_quick_panel = output_mode == 'quick_panel'
+        if show_in_quick_panel:
+            if modifier_keys.get('primary'):
+                side_by_side = True
+            self._show_references_in_quick_panel(word, session, response, side_by_side, force_group, group, position)
```

Comment:
this part was a bit hard to look a and understand, but I guess it is necessary if we want to support `primary` and `shift`